mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala

```diff
@@ -0,0 +1,41 @@
+package scala.meta.internal.mtags
+
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.concurrent.TrieMap
+
+/**
+ * This class is a wrapper around TrieMap that provides atomic updateWith
+ */
+final class AtomicTrieMap[K, V] {
+  private val trieMap = new TrieMap[K, V]()
+  private val concurrentMap = new ConcurrentHashMap[K, V]
+
+  def get(key: K): Option[V] = trieMap.get(key)
+
+  def contains(key: K): Boolean = trieMap.contains(key)
+
+  def updateWith(key: K)(remappingFunc: Option[V] => Option[V]): Unit = {
+    val computeFunction = new java.util.function.BiFunction[K, V, V] {
+      override def apply(k: K, v: V): V = {
+        trieMap.get(key) match {
```

Comment:
Not sure, but maybe if inside this function we referenced `k` instead of `key` (so input of `computeFunction` rather than the `updateWith` method) we could then extract it into a constant defined on the `AtomicTrieMap` class level so that there's no need to allocate a new one each time an `updateWith` is called? :thinking:

Reply:
I don't see its possible, because we need to pass it `remappingFunc`.
We could change `trieMap` type to `TrieMap[K, Set[V]]`, since we only use it like this and add a `append` function, that would look the way you suggested 🤔

Reply:
Sorry, you're absolutely right, somehow I missed the `remappingFunc`. :+1: